cache/redis_client.go

```diff
@@ -0,0 +1,463 @@
+	// -1 disables idle timeout check.
+	IdleTimeout time.Duration `yaml:"idle_timeout"`
+
+	// Connection age at which client retires (closes) the connection.
+	// Default 0 is to not close aged connections.
+	MaxConnAge time.Duration `yaml:"max_conn_age"`
+
+	// MaxAsyncConcurrency specifies the maximum number of SetAsync goroutines.
+	MaxAsyncConcurrency int `yaml:"max_async_concurrency"`
+
+	// MaxAsyncBufferSize specifies the queue buffer size for SetAsync operations.
+	MaxAsyncBufferSize int `yaml:"max_async_buffer_size"`
+
+	// MaxGetMultiConcurrency specifies the maximum number of concurrent GetMulti() operations.
+	// If set to 0, concurrency is unlimited.
+	MaxGetMultiConcurrency int `yaml:"max_get_multi_concurrency"`
+
+	// MaxItemSize specifies the maximum size of an item stored in redis.
+	// Items bigger than MaxItemSize are skipped.
+	// If set to 0, no maximum size is enforced.
+	MaxItemSize flagext.Bytes `yaml:"max_item_size"`
+
+	// GetMultiBatchSize specifies the maximum size per batch for mget.
+	GetMultiBatchSize int `yaml:"get_multi_batch_size"`
+
+	// TLSEnabled enable tls for redis connection.
+	TLSEnabled bool `yaml:"tls_enabled"`
+
+	// TLSConfig to use to connect to the redis server.
+	TLSConfig TLSConfig `yaml:"tls_config"`
+}
+
+func (c *RedisClientConfig) validate() error {
+	if c.Endpoint == "" {
+		return errMemcachedConfigNoEndpoint
+	}
+	// Set async only available when MaxAsyncConcurrency > 0.
+	if c.MaxAsyncConcurrency <= 0 {
+		return errRedisMaxAsyncConcurrencyNotPositive
+	}
+	return nil
+}
+
+func (c *RedisClientConfig) getAddresses() ([]string, error) {
+	addresses := strings.Split(c.Endpoint, ",")
+	// Handle single configuration endpoint which resolves multiple nodes.
+	if len(addresses) == 1 {
+		host, port, err := net.SplitHostPort(addresses[0])
+		if err != nil {
+			return nil, err
+		}
+		addrs, err := net.LookupHost(host)
+		if err != nil {
+			return nil, err
+		}
+		if len(addrs) > 1 {
+			addresses = nil
+			for _, addr := range addrs {
+				addresses = append(addresses, net.JoinHostPort(addr, port))
+			}
+		}
+	}
+	return addresses, nil
+}
+
+type redisClient struct {
+	redis.UniversalClient
+	config RedisClientConfig
+
+	// Channel used to notify internal goroutines when they should quit.
+	stop chan struct{}
+
+	// Channel used to enqueue async operations.
+	asyncQueue chan func()
+
+	// getMultiGate used to enforce the max number of concurrent GetMulti() operations.
+	getMultiGate gate.Gate
+
+	// Wait group used to wait all workers on stopping.
+	workers sync.WaitGroup
+
+	logger     log.Logger
+	operations *prometheus.CounterVec
+	failures   *prometheus.CounterVec
+	skipped    *prometheus.CounterVec
+	duration   *prometheus.HistogramVec
+	dataSize   *prometheus.HistogramVec
+}
+
+// NewRedisClientWithConfig makes a new RedisClient.
+func NewRedisClientWithConfig(logger log.Logger, name string, config RedisClientConfig, reg prometheus.Registerer) (RemoteCacheClient, error) {
```

Comment:
[nit] There's no other alternative than passing a config.

```suggestion
func NewRedisClient(logger log.Logger, name string, config RedisClientConfig, reg prometheus.Registerer) (RemoteCacheClient, error) {
```